File: backend/app/schemas/task.py
"""任务相关模式"""
from pydantic import BaseModel, Field, field_validator
from typing import Optional, List
from datetime import date, datetime
from decimal import Decimal
from app.models.task import TaskStatus, TaskPriority


class TaskBase(BaseModel):
    """任务基础模式"""
    title: str = Field(..., min_length=1, max_length=200, description="任务标题")
    description: Optional[str] = Field(None, description="任务描述")
    project_id: Optional[int] = Field(None, description="项目ID")
    estimated_man_days: Decimal = Field(..., ge=0, description="拟投入人天")
    required_skills: Optional[str] = Field(None, description="所需技能（JSON格式或逗号分隔）")
    deadline: Optional[date] = Field(None, description="截止时间")
    priority: str = Field(TaskPriority.P2.value, description="优先级：P0/P1/P2，默认P2")

    @field_validator("priority")
    @classmethod
    def validate_priority(cls, v: str) -> str:
        allowed = [p.value for p in TaskPriority]
        if v not in allowed:
            raise ValueError(f"优先级必须为 {allowed} 之一")
        return v


class TaskCreate(TaskBase):
    """创建任务请求"""
    pass


class TaskUpdate(BaseModel):
    """更新任务请求"""
    title: Optional[str] = Field(None, min_length=1, max_length=200, description="任务标题")
    description: Optional[str] = Field(None, description="任务描述")
    project_id: Optional[int] = Field(None, description="项目ID")
    estimated_man_days: Optional[Decimal] = Field(None, ge=0, description="拟投入人天")
    required_skills: Optional[str] = Field(None, description="所需技能")
    deadline: Optional[date] = Field(None, description="截止时间")
    is_pinned: Optional[bool] = Field(None, description="是否置顶（用于优先级）")
    priority: Optional[str] = Field(None, description="优先级：P0/P1/P2（仅草稿/已发布状态可修改）")

    @field_validator("priority")
    @classmethod
    def validate_priority(cls, v: Optional[str]) -> Optional[str]:
        if v is None:
            return v
        allowed = [p.value for p in TaskPriority]
        if v not in allowed:
            raise ValueError(f"优先级必须为 {allowed} 之一")
        return v


class TaskPublish(BaseModel):
    """发布任务请求"""
    pass


class TaskClaim(BaseModel):
    """认领任务请求"""
    pass


class TaskAssign(BaseModel):
    """派发任务请求"""
    assignee_id: int = Field(..., description="被派发的开发人员ID")


class TaskEvaluate(BaseModel):
    """评估任务请求"""
    accept: bool = Field(..., description="是否接受任务")


class TaskSubmit(BaseModel):
    """提交任务请求"""
    actual_man_days: Decimal = Field(..., gt=0, description="实际投入人天")


class TaskConfirm(BaseModel):
    """确认任务请求"""
    pass


class TaskReject(BaseModel):
    """退回已提交任务请求"""
    reason: str = Field(..., min_length=1, max_length=500, description="退回原因")


class TaskPin(BaseModel):
    """置顶任务请求"""
    is_pinned: bool = Field(..., description="是否置顶")


class TaskScheduleInfo(BaseModel):
    """任务排期信息（嵌入 TaskResponse）"""
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    is_pinned: bool = False
    is_concurrent: bool = False
    concurrent_with: Optional[int] = None

    class Config:
        from_attributes = True


class TaskResponse(BaseModel):
    """任务响应"""
    id: int
    title: str
    description: Optional[str]
    status: str
    project_id: Optional[int]
    creator_id: int
    assignee_id: Optional[int]
    estimated_man_days: Decimal
    actual_man_days: Optional[Decimal]
    required_skills: Optional[str]
    deadline: Optional[date]
    is_pinned: bool
    rejection_reason: Optional[str] = None
    priority: str = TaskPriority.P2.value
    priority_multiplier: Decimal = Decimal("1.00")
    schedule: Optional[TaskScheduleInfo] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class TaskDetailResponse(TaskResponse):
    """任务详情响应（包含关联信息）"""
    creator_name: Optional[str] = None
    assignee_name: Optional[str] = None
    project_name: Optional[str] = None


class TaskListItemResponse(TaskResponse):
    """任务列表项响应（包含关联信息）"""
    creator_name: Optional[str] = None
    assignee_name: Optional[str] = None
    project_name: Optional[str] = None


class TaskListResponse(BaseModel):
    """任务列表响应"""
    total: int
    items: List[TaskListItemResponse]


class TaskFilterParams(BaseModel):
    """任务筛选参数"""
    status: Optional[TaskStatus] = None
    statuses: Optional[List[TaskStatus]] = None
    project_id: Optional[int] = None
    project_ids: Optional[List[int]] = None
    creator_id: Optional[int] = None
    creator_ids: Optional[List[int]] = None
    assignee_id: Optional[int] = None
    assignee_ids: Optional[List[int]] = None
    keyword: Optional[str] = None
    required_skills: Optional[str] = None  # 所需技能（逗号分隔）
    priority: Optional[str] = None  # 优先级筛选
    page: int = Field(1, ge=1)
    page_size: int = Field(10, ge=1, le=100)


# ---- 并发排期相关 Schema ----

class SetConcurrentRequest(BaseModel):
    """设置并发请求"""
    concurrent_with_task_id: int = Field(..., description="基准任务ID（与哪个任务并发）")


class ExceededUser(BaseModel):
    """并发超限的用户信息"""
    user_id: int
    name: str
    current_concurrent: int
    limit: int


class AffectedSchedule(BaseModel):
    """受影响的排期（串行任务前移）"""
    task_id: int
    task_title: str
    old_scheduled_start: Optional[date]
    new_scheduled_start: Optional[date]


class ConcurrentCheckResponse(BaseModel):
    """并发预检响应"""
    can_set_concurrent: bool
    exceeded_users: List[ExceededUser] = []
    affected_schedules: List[AffectedSchedule] = []


# ---- 个人日程排期 Schema ----

class UserScheduleItem(BaseModel):
    """个人排期条目"""
    task_id: int
    task_title: str
    priority: str
    scheduled_start: Optional[date]
    scheduled_end: Optional[date]
    status: str
    estimated_days: float
    is_concurrent: bool
    concurrent_with: Optional[int]
    role: str  # "assignee" or "collaborator"
    is_pinned: bool = False


class UserScheduleResponse(BaseModel):
    """个人排期响应"""
    schedule: List[UserScheduleItem]


# ---- 项目排期 Schema ----

class ProjectScheduleItem(BaseModel):
    """项目排期条目（供甘特图使用）"""
    task_id: int
    task_title: str
    priority: str
    assignee_id: Optional[int]
    assignee_name: Optional[str]
    scheduled_start: Optional[date]
    scheduled_end: Optional[date]
    status: str
    estimated_man_days: float
    actual_man_days: Optional[float]
    is_concurrent: bool
    deadline: Optional[date]


class ProjectScheduleResponse(BaseModel):
    """项目排期响应"""
    project_id: int
    project_name: str
    tasks: List[ProjectScheduleItem]


# ---- 任务配合人相关 Schema ----

class CollaboratorAdd(BaseModel):
    """添加配合人请求"""
    user_id: int = Field(..., description="配合人用户ID")
    allocated_man_days: Decimal = Field(..., gt=0, description="分配人天数，必须大于0")


class CollaboratorUpdate(BaseModel):
    """更新配合人人天请求"""
    allocated_man_days: Decimal = Field(..., gt=0, description="更新后的分配人天数")


class CollaboratorResponse(BaseModel):
    """配合人响应"""
    id: int
    task_id: int
    user_id: int
    user_name: Optional[str] = None
    user_full_name: Optional[str] = None
    allocated_man_days: Decimal
    scheduled_start: Optional[date] = None
    scheduled_end: Optional[date] = None
    created_at: datetime

    class Config:
        from_attributes = True

File: frontend/src/api/task.ts
/**
 * 任务管理API
 */
import request from './index'

// 优先级常量
export const PRIORITY_OPTIONS = [
  { value: 'P2', label: 'P2 常规', color: '#909399', multiplier: 1.0 },
  { value: 'P1', label: 'P1 较紧急', color: '#E6A23C', multiplier: 1.1 },
  { value: 'P0', label: 'P0 最紧急', color: '#F56C6C', multiplier: 1.2 },
] as const

export type TaskPriority = 'P0' | 'P1' | 'P2'

export const PRIORITY_MULTIPLIER: Record<TaskPriority, number> = {
  P0: 1.2,
  P1: 1.1,
  P2: 1.0,
}

export interface TaskScheduleInfo {
  start_date?: string
  end_date?: string
  is_pinned: boolean
  is_concurrent: boolean
  concurrent_with?: number
}

export interface Task {
  id: number
  title: string
  description?: string
  status: string
  project_id?: number
  creator_id: number
  assignee_id?: number
  estimated_man_days: number
  actual_man_days?: number
  required_skills?: string
  deadline?: string
  is_pinned: boolean
  rejection_reason?: string
  priority: TaskPriority
  priority_multiplier: number
  schedule?: TaskScheduleInfo
  created_at: string
  updated_at: string
}

export interface TaskDetail extends Task {
  creator_name?: string
  assignee_name?: string
  project_name?: string
}

export interface TaskCreate {
  title: string
  description?: string
  project_id?: number
  estimated_man_days: number
  required_skills?: string
  deadline?: string
  priority?: TaskPriority
}

export interface TaskUpdate {
  title?: string
  description?: string
  project_id?: number
  estimated_man_days?: number
  required_skills?: string
  deadline?: string
  priority?: TaskPriority
}

export interface TaskListResponse {
  total: number
  items: Task[]
}

export interface TaskFilterParams {
  status?: string
  project_id?: number
  creator_id?: number
  assignee_id?: number
  keyword?: string
  priority?: TaskPriority
  page?: number
  page_size?: number
}

export interface TaskAssign {
  assignee_id: number
}

export interface TaskEvaluate {
  accept: boolean
}

export interface TaskSubmit {
  actual_man_days: number
}

export interface TaskPin {
  is_pinned: boolean
}

// ---- 并发排期相关类型 ----

export interface SetConcurrentRequest {
  concurrent_with_task_id: number
}

export interface ExceededUser {
  user_id: number
  name: string
  current_concurrent: number
  limit: number
}

export interface AffectedSchedule {
  task_id: number
  task_title: string
  old_scheduled_start?: string
  new_scheduled_start?: string
}

export interface ConcurrentCheckResponse {
  can_set_concurrent: boolean
  exceeded_users: ExceededUser[]
  affected_schedules: AffectedSchedule[]
}

// ---- 个人日程类型 ----

export interface UserScheduleItem {
  task_id: number
  task_title: string
  priority: TaskPriority
  scheduled_start?: string
  scheduled_end?: string
  status: string
  estimated_days: number
  is_concurrent: boolean
  concurrent_with?: number
  role: 'assignee' | 'collaborator'
  is_pinned: boolean
}

export interface UserScheduleResponse {
  schedule: UserScheduleItem[]
}

// ---- 项目排期类型 ----

export interface ProjectScheduleTask {
  task_id: number
  task_title: string
  priority: TaskPriority
  priority_multiplier: number
  assignee_id?: number
  assignee_name?: string
  scheduled_start?: string
  scheduled_end?: string
  status: string
  estimated_man_days: number
  actual_man_days?: number
  is_concurrent: boolean
  concurrent_with?: number
  deadline?: string
}

export interface ProjectScheduleResponse {
  project_id: number
  project_name: string
  tasks: ProjectScheduleTask[]
}

/**
 * 创建任务
 */
export function createTask(data: TaskCreate): Promise<Task> {
  return request.post('/api/v1/tasks', data)
}

/**
 * 获取任务列表
 */
export function getTasks(params?: TaskFilterParams): Promise<TaskListResponse> {
  return request.get('/api/v1/tasks', { params })
}

/**
 * 获取任务详情
 */
export function getTask(taskId: number): Promise<TaskDetail> {
  return request.get(`/api/v1/tasks/${taskId}`)
}

/**
 * 更新任务
 */
export function updateTask(taskId: number, data: TaskUpdate): Promise<Task> {
  return request.put(`/api/v1/tasks/${taskId}`, data)
}

/**
 * 删除任务
 */
export function deleteTask(taskId: number): Promise<void> {
  return request.delete(`/api/v1/tasks/${taskId}`)
}

export interface MarketplaceTask {
  id: number
  title: string
  description?: string
  status: string
  project_id?: number
  project_name?: string
  creator_id: number
  creator_name?: string
  estimated_man_days: number
  required_skills?: string
  deadline?: string
  created_at?: string
  priority: TaskPriority
  priority_multiplier: number
}

export interface MarketplaceResponse {
  tasks: MarketplaceTask[]
  total: number
  page: number
  page_size: number
}

/**
 * 获取任务集市数据
 */
export function getMarketplaceTasks(params?: {
  status?: string
  project_id?: number
  keyword?: string
  required_skills?: string
  priority?: TaskPriority
  recommend?: boolean
  page?: number
  page_size?: number
}): Promise<MarketplaceResponse> {
  return request.get('/api/v1/tasks/marketplace', { params })
}

/**
 * 发布任务
 */
export function publishTask(taskId: number): Promise<Task> {
  return request.post(`/api/v1/tasks/${taskId}/publish`)
}

/**
 * 认领人退回 / 发起人收回已认领任务，回到"已发布"状态
 */
export function returnTask(taskId: number): Promise<Task> {
  return request.post(`/api/v1/tasks/${taskId}/return`)
}

/**
 * 将已发布任务退回草稿
 */
export function revertTaskToDraft(taskId: number): Promise<Task> {
  return request.post(`/api/v1/tasks/${taskId}/revert-draft`)
}

// ---- 任务配合人 ----

export interface Collaborator {
  id: number
  task_id: number
  user_id: number
  user_name?: string
  user_full_name?: string
  allocated_man_days: number
  scheduled_start?: string
  scheduled_end?: string
  created_at: string
}

/**
 * 获取任务配合人列表
 */
export function getCollaborators(taskId: number): Promise<Collaborator[]> {
  return request.get(`/api/v1/tasks/${taskId}/collaborators`)
}

/**
 * 添加配合人
 */
export function addCollaborator(
  taskId: number,
  data: { user_id: number; allocated_man_days: number }
): Promise<Collaborator> {
  return request.post(`/api/v1/tasks/${taskId}/collaborators`, data)
}

/**
 * 更新配合人分配人天
 */
export function updateCollaborator(
  taskId: number,
  collaboratorUserId: number,
  data: { allocated_man_days: number }
): Promise<Collaborator> {
  return request.put(`/api/v1/tasks/${taskId}/collaborators/${collaboratorUserId}`, data)
}

/**
 * 移除配合人
 */
export function removeCollaborator(taskId: number, collaboratorUserId: number): Promise<void> {
  return request.delete(`/api/v1/tasks/${taskId}/collaborators/${collaboratorUserId}`)
}

/**
 * 认领任务
 */
export function claimTask(taskId: number): Promise<Task> {
  return request.post(`/api/v1/tasks/${taskId}/claim`)
}

/**
 * 派发任务
 */
export function assignTask(taskId: number, data: TaskAssign): Promise<Task> {
  return request.post(`/api/v1/tasks/${taskId}/assign`, data)
}

/**
 * 评估任务（接受或拒绝）
 */
export function evaluateTask(taskId: number, data: TaskEvaluate): Promise<Task> {
  return request.post(`/api/v1/tasks/${taskId}/evaluate`, data)
}

/**
 * 开始任务
 */
export function startTask(taskId: number): Promise<Task> {
  return request.post(`/api/v1/tasks/${taskId}/start`)
}

/**
 * 提交任务
 */
export function submitTask(taskId: number, data: TaskSubmit): Promise<Task> {
  return request.post(`/api/v1/tasks/${taskId}/submit`, data)
}

/**
 * 确认任务
 */
export function confirmTask(taskId: number): Promise<Task> {
  return request.post(`/api/v1/tasks/${taskId}/confirm`)
}

/**
 * 退回已提交任务（发起人/PM），附带退回原因，任务回到"进行中"状态
 */
export function rejectTask(taskId: number, reason: string): Promise<Task> {
  return request.post(`/api/v1/tasks/${taskId}/reject`, { reason })
}

/**
 * 置顶/取消置顶任务
 */
export function pinTask(taskId: number, data: TaskPin): Promise<Task> {
  return request.post(`/api/v1/tasks/${taskId}/pin`, data)
}

/**
 * 获取任务排期
 */
export function getTaskSchedule(taskId: number): Promise<any> {
  return request.get(`/api/v1/tasks/${taskId}/schedule`)
}

// ---- 并发排期接口 ----

/**
 * 并发预检（不执行实际设置）
 */
export function checkConcurrent(
  taskId: number,
  concurrentWithTaskId: number
): Promise<ConcurrentCheckResponse> {
  return request.get(`/api/v1/tasks/${taskId}/concurrent-check`, {
    params: { concurrent_with_task_id: concurrentWithTaskId },
  })
}

/**
 * 设置任务为并发
 */
export function setConcurrent(
  taskId: number,
  data: SetConcurrentRequest
): Promise<any> {
  return request.post(`/api/v1/tasks/${taskId}/set-concurrent`, data)
}

/**
 * 取消任务并发状态
 */
export function unsetConcurrent(taskId: number): Promise<any> {
  return request.delete(`/api/v1/tasks/${taskId}/set-concurrent`)
}

/**
 * 获取当前用户的完整排期
 */
export function getMySchedule(params?: {
  start_date?: string
  end_date?: string
}): Promise<UserScheduleResponse> {
  return request.get('/api/v1/tasks/me/schedule', { params })
}

/**
 * 手动触发当前用户排期重算（任务提前完成后使用）
 */
export function recalculateMySchedule(): Promise<{ success: boolean; message: string }> {
  return request.post('/api/v1/tasks/me/recalculate-schedule')
}

/**
 * 获取项目排期（甘特图数据）
 */
export function getProjectSchedule(projectId: number): Promise<ProjectScheduleResponse> {
  return request.get(`/api/v1/projects/${projectId}/schedule`)
}

// ── 任务留言 ─────────────────────────────────────────────────────────────────

export interface TaskComment {
  id: number
  task_id: number
  user_id: number
  user_name: string
  user_full_name?: string
  content: string
  created_at: string
  updated_at: string
}

export interface TaskCommentListResponse {
  total: number
  items: TaskComment[]
}

/**
 * 获取任务留言列表
 */
export function getTaskComments(taskId: number): Promise<TaskCommentListResponse> {
  return request.get(`/api/v1/tasks/${taskId}/comments`)
}

/**
 * 发表任务留言
 */
export function createTaskComment(taskId: number, content: string): Promise<TaskComment> {
  return request.post(`/api/v1/tasks/${taskId}/comments`, { content })
}

/**
 * 编辑留言
 */
export function updateTaskComment(commentId: number, content: string): Promise<TaskComment> {
  return request.put(`/api/v1/tasks/comments/${commentId}`, { content })
}

/**
 * 删除留言
 */
export function deleteTaskComment(commentId: number): Promise<void> {
  return request.delete(`/api/v1/tasks/comments/${commentId}`)
}
